gadgets/trace_udns/program.bpf.c

```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+/* Copyright (c) 2024 The Inspektor Gadget authors */
+
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+
+#include <gadget/buffer.h>
+#include <gadget/macros.h>
+#include <gadget/mntns_filter.h>
+#include <gadget/types.h>
+
+#define AF_INET 2
+#define AF_INET6 10
+
+enum op {
+	getaddrinfo,
+	gethostbyname,
+	gethostbyname2,
+};
+
+// Max DNS name length: 255
+// https://datatracker.ietf.org/doc/html/rfc1034#section-3.1
+#define MAX_DNS_NAME 255
+
+struct event {
+	gadget_mntns_id mntns_id;
+	__u32 pid;
+	__u32 tid;
+	__u8 comm[TASK_COMM_LEN];
+	enum op operation;
+	__u8 name[MAX_DNS_NAME];
+	__u8 service[MAX_DNS_NAME];
+	struct gadget_l3endpoint_t endpoint;
+};
+
+GADGET_TRACER_MAP(events, 1024 * 256);
+
+GADGET_TRACER(udns, events, event);
+
+// struct addrinfo comes from netdb.h in glibc or musl:
+// https://sourceware.org/git/?p=glibc.git;a=blob;f=resolv/netdb.h;hb=ded2e0753e9c46debeb2e0d26c5e560d2581d314#l565
+// https://git.etalabs.net/cgit/musl/tree/include/netdb.h#n16
+struct addrinfo {
+	int ai_flags;
+	int ai_family;
+	int ai_socktype;
+	int ai_protocol;
+	__u32 ai_addrlen;
+	struct sockaddr *ai_addr;
+	char *ai_canonname;
+	struct addrinfo *ai_next;
+};
+
+struct lookup {
+	const char *node;
+	const char *service;
+	const void *hints;
+	const struct addrinfo **res;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, 10240);
+	__type(key, u32); // tid
+	__type(value, struct lookup);
+} lookups SEC(".maps");
+
+SEC("uprobe/libc:getaddrinfo")
+int BPF_KPROBE(getaddrinfo_e, const char *node, const char *service,
```

Comment:
weak suggestion: you can also use `BPF_UPROBE`, and the two macros are exactly the same.